### PR TITLE
Add annotation labels for 3D plots

### PR DESCRIPTION
**Pull request by t0mdavid-m**
This PR adds annotation labels for 3D plots. It can be tested easily using the `PeakMap` example jupyter notebook with the following sample `DataFrame`:

```
annotation_data = pd.DataFrame(
    {
        'leftWidth' : [212],
        'rightWidth' : [237],
        'IM_leftWidth' : [266],
        'IM_rightWidth' : [272],
        'color' : ['blue'],
        'name' : ['test'],
    }
)
```

While I was at it I also enabled the colors for the 2D visualizations (optional). I noticed that the rectangular 2D annotation was shown behind the plot data and made sure that it is displayed in the foreground.

I also noticed that  annotation color and label for 3D plots is set using the input `DataFrame`. I propose moving these annotations to the `annotation_data` `DataFrame` to simplify the user experience.

### CODE REVIEW

File: pyopenms_viz/_core.py [hunk 1]
```diff
@@ -1258,6 +1259,35 @@ def _add_box_boundaries(self, annotation_data):
...
+        
+        # Contains tuple of coordinates + text + color (x, y, z, t, c)
+        annotations_3d = []
+        for _, feature in annotation_data.iterrows():
+            x0 = feature["leftWidth"]
+            x1 = feature["rightWidth"]
+            y0 = feature["IM_leftWidth"]
+            y1 = feature["IM_rightWidth"]
+            t = feature["name"]
+            c = feature["color"]
```
Comment (jcharkow) [inline pyopenms_viz/_core.py:1275]: Suggestion not to hard code these column names, include as kwargs for function. Possibly in the future will support more column names.

File: pyopenms_viz/_matplotlib/core.py [hunk 2]
```diff
@@ -732,7 +757,6 @@ def _add_box_boundaries(self, annotation_data):
                 loc=matplotlibLegendLoc,
                 title=self.feature_config.legend.title,
                 prop={"size": self.feature_config.legend.fontsize},
-                bbox_to_anchor=self.feature_config.legend.bbox_to_anchor,
```
Comment (jcharkow) [inline pyopenms_viz/_matplotlib/core.py:735]: why is this line removed? I think this is needed for 2D legend placement

File: pyopenms_viz/_plotly/core.py [hunk 3]
```diff
@@ -793,26 +824,34 @@ def _add_box_boundaries(self, annotation_data, **kwargs):
...
-                    opacity=0.5,
                     line=dict(
                         color=color,
                         width=self.feature_config.line_width,
-                        dash=bokeh_line_dash_mapper(
-                            self.feature_config.line_type, "plotly"
-                        ),
+                        dash=bokeh_line_dash_mapper(self.feature_config.line_type, "plotly"),
                     ),
+                    showlegend=True,
```
Comment (jcharkow) [inline pyopenms_viz/_plotly/core.py:854]: should legend showing be configurable?

**Review by jcharkow — APPROVED** [review]
Looks good just some minor suggestions. @singjc is more familiar with this so I would wait for his input before merging. I have not tested this out so assuming it does not break the existing 2D annotations.

File: pyopenms_viz/_core.py (hunk 1 above)
Comment (t0mdavid-m) [reply pyopenms_viz/_core.py]: Done!

File: pyopenms_viz/_matplotlib/core.py (hunk 2 above)
Comment (t0mdavid-m) [reply pyopenms_viz/_matplotlib/core.py:735]: Sorry - I forgot to mention this. There seems to be a weird interaction between `bbox_to_anchor` and `tight_layout()`, where the plot gets resized when the legend is added. With the legend enabled not only the plot area shrinks (expected as the legend needs space) but also the entire figure gets resized in a strange way (unexpected).

If I disable `tight_layout()` in `MATPLOTLIBPlot.show_default()` or remove the `bbox_to_anchor` parameter the issues disappear. 

How would you recommend to handle this issue?

File: pyopenms_viz/_plotly/core.py (hunk 3 above)
Comment (t0mdavid-m) [reply pyopenms_viz/_plotly/core.py:853]: I suppose we could make it optional for 3D plots as we have the in-plot annotation labels with this PR. For 2D-plots, the legend is currently the only way to pick a location so I am not sure how useful that would be.

We could add the same functionality but I am unsure how to best choose a good location for the in-plot annotation labels. Maybe we could allow for additional (x, y)-label-locations  in `annotation_data` and let the user decide? What do you think?

**Review by singjc — APPROVED** [review]
I tested and updated the `manuscript_supp_config_gallery` with the text annotations. It seems to work nicely with matplotlib and plotly. I did add a commit to allow adjusting the font size of the text annotations.

Other than Josh's comments/suggestions, I think everything looks good.

File: pyopenms_viz/_plotly/core.py (hunk 3 above)
Comment (singjc) [reply pyopenms_viz/_plotly/core.py:853]: Is this for adding in-plot text annotations 2D plots, that requires creating a dummy legend?

File: pyopenms_viz/_matplotlib/core.py (hunk 2 above)
Comment (singjc) [reply pyopenms_viz/_matplotlib/core.py:735]: If you disable `tight_layout` in `MATPLOTLIBPlot.show_default()` and keep `bbox_to_anchor`, after the plot get's generated, if you call `plt.tight_layout()`, does it still cause strange resizing?

File: pyopenms_viz/_plotly/core.py (hunk 3 above)
Comment (jcharkow) [reply pyopenms_viz/_plotly/core.py:853]: Any updates on this? I am not too familiar with Plotly so not sure what is best

**Comment by jcharkow** [comment]
Hopefully we can merge soon?

File: pyopenms_viz/_plotly/core.py (hunk 3 above)
Comment (t0mdavid-m) [reply pyopenms_viz/_plotly/core.py:853]: > Is this for adding in-plot text annotations 2D plots, that requires creating a dummy legend?

A dummy legend is not necessary for in-plot annotations afaik. I was just wondering how useful making the legend optional would be if there is no alternative. Thus, I suggested adding annotation labels to 2D-Plots as well but am unsure about a good general placement for the label.

Comment (t0mdavid-m) [reply pyopenms_viz/_plotly/core.py:853]: I suppose we could require a set of coordinates for the user. This could be an optional feature for the 3D-Plots as well.

File: pyopenms_viz/_matplotlib/core.py (hunk 2 above)
Comment (t0mdavid-m) [reply pyopenms_viz/_matplotlib/core.py]: No this did not resolve this issue on its own. Calling `plt.tight_layout()` after the fact seemed to not have any effect. Adjusting the anchoring of the `bbox`  seems to result in nice plots (and should be more stable from what I can tell):
image

**Comment by singjc** [comment]
> Hopefully we can merge soon?

I think it is ready to merge.

**Comment by singjc** [comment]
> I agree @singjc .
> 
> Did you also see my reply to the comment regarding the configurable legend? @jcharkow @singjc
> 
> If we decide to implement it I suppose it could also be done in a separate PR.

I think your suggestion for having additional x-y coordinate locations for positioning the labels could be an option? Then the legend could be set as configurable to display a legend or not. We could add it to this PR if it's not too much/won't take too long, or we could implement in another PR.

I think we would like to merge this PR relatively soon, since it is used in the manuscript.

**Comment by t0mdavid-m** [comment]
Agree that this PR should be merged asap. I will put this on my ToDo list and open a new PR when the changes are ready.

**Comment by jcharkow** [comment]
Will merge now
